### ISSUE TITLE
Are AnnData input files officially supported?

### ISSUE THREAD

**Issue by tomsing1** [broadinstitute/CellBender#166]
After closing #88 it seems that AnnData objects, serialized as `h5ad` files, are now valid inputs for CellBender?
The `README.rst` states `At the moment, only the count matrices produced by the CellRanger count pipeline is supported` - but perhaps that's out of date?

In case AnnData objects _are_ supported, are there any known limitations? Specifically, are outputs from [alevin-fry](https://github.com/COMBINE-lab/alevin-fry) compatible with CellBender? I believe by default `alevin-fry` only outputs counts for droplets with >= 10 counts. Would that still be valid input for CellBender?

**Comment by sjfleming** [comment]
Hi @tomsing1 , yes you are correct.  That readme is out of date!

AnnData objects are supported as inputs as of #88 .  Any sort of integer count matrix is acceptable, including alevin-fry, starsolo, cellranger, etc.  The only thing cellbender needs, as you allude to, is a lot of examples of empty droplets that do not have cells.

I do not have hands-on experience with `alevin-fry` myself.  If it only outputs counts for droplets with >= 10 counts, then that should be sufficient in most cases for the following reason: most experiments I see have >= 10 counts in the empty droplets.  If you end up in a situation where the empty droplets really have < 10 counts, then you're very nearly in the regime where the dataset is so clean to begin with that you might not really need `cellbender` to clean up the dataset in the first place.  So hopefully those raw `alevin-fry` count matrices would work fine.

One question I'd have about the `alevin-fry` data structure would be that, looking at their documentation, I see they quantify "unspliced, spliced and ambiguous genes".  I wonder in particular about the category called "ambiguous genes".  I'm guessing that does not end up in the count matrix you would be feeding to `cellbender`, is that right?  Would you keep both spliced and unspliced?  (Summing them, or acting as though they are separate features?)  Or just use the spliced count matrix?

**Comment by tomsing1** [comment]
Thanks a lot for confirming @sjfleming - that's awesome! 

Alevin-fry  generates the final count matrix in two steps: as you pointed out, initially it provides 3 types of counts per gene: unspliced (`U`), spliced (`S`) and ambiguous (e.g. intronic) `(A)`. 

The user can then choose which of them to use for downstream analysis, e.g. with the [pyroe](https://github.com/COMBINE-lab/pyroe#processing-alevin-fry-quantification-result) python module or the `loadFry()` function from the [fishpond](https://bioconductor.org/packages/release/bioc/html/fishpond.html) R package. 

These helper functions allow users to either choose a preset mode (`scRNA`, `snRNA` or `velocity`) or define their own combination of the U, S and A counts. 

- The `snRNA` mode will sum all three types of counts (U+S+A) per gene and droplet
- The `scRNA` mode outputs S+A per gene and droplet
- `velocity` mode returns two assay matrices: 
   1. `spliced` (S+A) and 
   2. `unspliced` (U) counts of each gene in each droplet. 

To mirror cellranger 7's behavior, which also counts intronic alignments, I am using `snRNA` (U+S+A) for _both_ single-cell and single-nuclei experiments right now.

**Comment by tomsing1** [comment]
Update: Rob Patro, one of the authors of alevin-fry and its [simpleaf wrapper](https://github.com/COMBINE-lab/simpleaf/releases/tag/v0.7.0) has exposed an optional `--min-reads` argument now, which allows users to control the filtering applied by the alevin-fry `quant` stage. This allows users to tune the output e.g. in the presence of different levels of noise by setting a cutoff other than >= 10.

**Comment by sjfleming** [comment]
Very cool!

Then yes, `cellbender` should work fine with this kind of anndata h5ad from `alevin-fry`.  Let me know if you run into any problems!

One other thing to note: the logic of the anndata loader is such that it will accept a count matrix in a couple different "locations" inside the h5ad file.  From the docstring 
```
The file needs to contain raw counts for all measured barcodes in the `.X` attribute 
or a `.layers[{'counts', 'spliced'}]` attribute.
```
The logic is actually set up such that a count matrix in `adata.layers['counts']` takes top priority, followed by `adata.layers['spliced']`, and then the fallback is to take the count matrix from `adata.X`.

**Comment by tomsing1** [comment]
Oh, that's great to know. Thanks a lot for pointing out the priority of the assay layers!

**Comment by sjfleming** [comment]
Closed by #88
